src/utils/component-versioning.js

```diff
@@ -101,11 +120,16 @@ const versionedHtml = componentInstance => (strings, ...args) => {
   // extract dependency info by looking at versions of component
   const [tagNames, versions] = extractDependencies(componentClass);
   // rewrite incoming array of static parts of template literals
-  // in such a way that tags of dependent components are versioned
+  // in such a way that tags of dependent components are versioned:
   let newStrings = strings;
+  // 1. rewriting proper, turning tags into versioned ones with funny initial brackets (see newTag(...) above)
   for (let i = 0, n = tagNames.length; i < n; i++) {
     newStrings = rewrite(newStrings, tagNames[i], versions[i]);
   }
+  // 2. finish rewriting by converting funny initial brackets back to standard ones
+  for (let i = 0, n = newStrings.length; i < n; i++) {
```

Comment:
Why 2 variables? I am sure you have a reason for that (maybe infinite loop and outofmemory at some point, because array is modified within the body, which COULD BE a potential risk). :)

This is simpler in my opinion:
``` for (let i = 0; i < newStrings.length; i++) {```

Reply:
This is a generic pattern for JS performance: cache the results of a computation that is not changed in a loop.

taking the length of a string is such a computation.

This idiom is used in all of my code...